integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+import java.sql.SQLException;
+import java.util.EnumSet;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.transaction.xa.XAException;
+import javax.transaction.xa.XAResource;
+import javax.transaction.xa.Xid;
+
+import io.helidon.integrations.jdbc.SQLRunnable;
+import io.helidon.integrations.jdbc.UncheckedSQLException;
+import io.helidon.integrations.jta.jdbc.ExceptionConverter.XARoutine;
+
+import static javax.transaction.xa.XAException.XAER_DUPID;
+import static javax.transaction.xa.XAException.XAER_INVAL;
+import static javax.transaction.xa.XAException.XAER_NOTA;
+import static javax.transaction.xa.XAException.XAER_PROTO;
+import static javax.transaction.xa.XAException.XAER_RMERR;
+import static javax.transaction.xa.XAException.XAER_RMFAIL;
+import static javax.transaction.xa.XAException.XA_RBROLLBACK;
+import static javax.transaction.xa.XAResource.TMENDRSCAN;
+import static javax.transaction.xa.XAResource.TMFAIL;
+import static javax.transaction.xa.XAResource.TMJOIN;
+import static javax.transaction.xa.XAResource.TMNOFLAGS;
+import static javax.transaction.xa.XAResource.TMONEPHASE;
+import static javax.transaction.xa.XAResource.TMRESUME;
+import static javax.transaction.xa.XAResource.TMSTARTRSCAN;
+import static javax.transaction.xa.XAResource.TMSUCCESS;
+import static javax.transaction.xa.XAResource.TMSUSPEND;
+import static javax.transaction.xa.XAResource.XA_OK;
+import static javax.transaction.xa.XAResource.XA_RDONLY;
+
+/**
+ * An {@link XAResource} that adapts an ordinary arbitrary {@link Connection} as much as possible to the {@link
+ * XAResource} contract.
+ *
+ * <p><strong>Note:</strong> instances of this class are lossless in the presence of one-phase commit operations and
+ * potentially lossy in the presence of two-phase commit operations.</p>
+ *
+ * <p>Instances of this class are safe for concurrent use by multiple threads.</p>
+ */
+final class LocalXAResource implements XAResource {
+
+
+    /*
+     * Static fields.
+     */
+
+
+    private static final Logger LOGGER = Logger.getLogger(LocalXAResource.class.getName());
+
+    private static final Xid[] EMPTY_XID_ARRAY = new Xid[0];
+
+    // package-protected for testing only.
+    static final ConcurrentMap<Xid, Association> ASSOCIATIONS = new ConcurrentHashMap<>();
+
+
+    /*
+     * Instance fields.
+     */
+
+
+    private final Function<? super Xid, ? extends Connection> connectionFunction;
+
+    private final ExceptionConverter exceptionConverter;
+
+
+    /*
+     * Constructors.
+     */
+
+
+    /**
+     * Creates a new {@link LocalXAResource}.
+     *
+     * @param connectionFunction a {@link Function} that accepts a {@link Xid} (supplied by the {@link #start(Xid, int)}
```

Comment:
Technically a connection doesn't have to be supplied at construction time, because it won't be used unless `start` is called.  So this implementation allows you to pass a `Function` that will be called by the `start` method, if and when it is called.